sbncode/CAFMaker/FillTrue.cxx

```diff
@@ -127,6 +130,190 @@ namespace caf {
+          truep.y = (truep.y*truep.p_nelec + ide->y*ide->numElectrons) / (truep.p_nelec + ide->numElectrons);
+          truep.z = (truep.z*truep.p_nelec + ide->z*ide->numElectrons) / (truep.p_nelec + ide->numElectrons);
+
+          truep.p_e += ide->energy;
+          truep.p_nelec += ide->numElectrons;
+        }
+
+        // sim::IDE's are saved after shift from SpaceCharge
+        //
+        // Correct the mean position for space charge to find the MCParticle trajectory match
+        geo::Point_t loc_sce {truep.x, truep.y, truep.z}; // perturbed by space charge
+        geo::Point_t loc_nosce; // corrected for space charge
+        if (sce && sce->EnableSimSpatialSCE()) {
+          loc_nosce = loc_sce + sce->GetCalPosOffsets(loc_sce, p.tpc);
+        }
+        else loc_nosce = loc_sce;
+
+        TVector3 loc_nosce_v(loc_nosce.x(), loc_nosce.y(), loc_nosce.z());
+        
+        // Match to MC-trajectory
+        TVector3 direction;
+        float closest_dist = -1.;
+        int traj_index = -1;
+        for (unsigned i_traj = 0; i_traj < particle.NumberTrajectoryPoints(); i_traj++) {
+          if (closest_dist < 0. || (particle.Position(i_traj).Vect() - loc_nosce_v).Mag() < closest_dist) {
+            direction = particle.Momentum(i_traj).Vect().Unit();
+            closest_dist = (particle.Position(i_traj).Vect() - loc_nosce_v).Mag();
+            traj_index = i_traj;
+          }
+        }
+
+        // residual range
+        truep.rr = 0;
+        if (traj_index >= 0) {
+          for (int i_traj = traj_index+1; i_traj < (int)particle.NumberTrajectoryPoints(); i_traj++) {
+            truep.rr += (particle.Position(i_traj).Vect() - particle.Position(i_traj-1).Vect()).Mag();
+          }
+
+          // Also account for the distance from the Hit point to the matched trajectory point
+          double hit_distance_along_particle = (loc_nosce_v - particle.Position(traj_index).Vect()).Dot(particle.Momentum(traj_index).Vect().Unit());
+          truep.rr += -hit_distance_along_particle;
+        }
+        else truep.rr = -1;
+
+        // pitch
+        //
+        // The true pitch is affected by the presence of space charge, so we cannot
+        // directly apply the true direction. We include the effect of space charge
+        // on the true pitch here
+        if (closest_dist >= 0. && direction.Mag() > 1e-4) {
+          geo::PlaneID plane(srtrack.producer, p.tpc, iplane);
+          float angletovert = geo->WireAngleToVertical(geo->View(plane), plane) - 0.5*::util::pi<>();
+          TVector3 loc_mdx_v = loc_nosce_v - direction * (geo->WirePitch(geo->View(plane)) / 2.);
+          TVector3 loc_pdx_v = loc_nosce_v + direction * (geo->WirePitch(geo->View(plane)) / 2.);
+
+          // Convert types for helper functions
+          geo::Point_t loc_mdx(loc_mdx_v.X(), loc_mdx_v.Y(), loc_mdx_v.Z());
+          geo::Point_t loc_pdx(loc_pdx_v.X(), loc_pdx_v.Y(), loc_pdx_v.Z());
```

Comment:
Assuming `direction` to be defined as `geo::Vector_t` instead of `TVector3` (you can use a `geo::vect::toVector()` on line 239, for example):
```suggestion
          geo::Point_t loc_mdx = loc_nosce - direction * plane.WirePitch() / 2.;
          geo::Point_t loc_pdx = loc_nosce + direction * plane.WirePitch() / 2.;
```
I pruned `angletovert` (it would have been `double const angletovert = plane.PhiZ();`) because it won't be needed, see below.

Reply:
This code copies the interfaces used in the Calorimetry code for calculating pitch. I'd rather keep the use of interfaces consistent. I've been burned before by code behavior changing in LArSoft services and I'd rather reduce the number of different interfaces I use overall.